### PR TITLE
URL shortner for dashboards

### PR DESCRIPTION
**Pull request by ttannis**
This PR addresses issue #4309 filed by @vylc. A URLShortLinkButton has been added to the dashboard, and the component has been refactored to be able to be shared in the dashboard view and explore view.

url_dashboard

### CODE REVIEW

File: superset/assets/javascripts/components/URLShortLinkButton.jsx [hunk 1]
```diff
@@ -25,20 +27,20 @@ export default class URLShortLinkButton extends React.Component {
   }
 
   getCopyUrl() {
-    const longUrl = getExploreLongUrl(this.props.latestQueryFormData);
+    const longUrl = (this.props.type === 'Dashboard') ? getDashboardLongUrl(this.props.context) : getExploreLongUrl(this.props.context);
```
[the remark this answers is not in the document]
Comment (ttannis) [reply superset/assets/javascripts/components/URLShortLinkButton.jsx]: Yeah, I noticed after uploading the effect on code coverage. I'll get to fixing that.

File: superset/assets/javascripts/dashboard/dashboardUtils.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+/* eslint camelcase: 0 */
+import URI from 'urijs';
+
+export function getURIDirectory(dashboard) {
```
[the remark this answers is not in the document]
Comment (ttannis) [reply superset/assets/javascripts/dashboard/dashboardUtils.js]: 👍

File: superset/assets/javascripts/dashboard/components/Dashboard.jsx [hunk 3]
```diff
@@ -153,6 +157,27 @@ class Dashboard extends React.PureComponent {
     return this.props.filters[sliceId];
   }
 
+  addHistory({ isReplace = false, title }) {
```
Comment (mistercrunch) [inline superset/assets/javascripts/dashboard/components/Dashboard.jsx]: I'm confused about this method:
* why using destructuring in the function parameters?
* why do we need to alter the browser history?
* `title` is never passed to it, won't that lead to clearing the title?

File: superset/assets/javascripts/components/URLShortLinkButton.jsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (mistercrunch) [reply superset/assets/javascripts/components/URLShortLinkButton.jsx]: Also not sure why `getDashboardLongUrl` is necessary in the first place, can we just use `window.location`? Assuming we want the current state of the dashboard (with current filters), `window.location` should be in sync at that point.

**Review by mistercrunch — CHANGES_REQUESTED** [review]
A few comments about making the `URLShortLinkButton` component as generic as possible, and let's clarify whether we need url manipulations.

**Comment by mistercrunch** [comment]
screen shot 2018-04-09 at 2 04 39 pm
!!!!

File: superset/assets/javascripts/components/URLShortLinkButton.jsx [hunk 4]
```diff
@@ -1,13 +1,14 @@
...
-import { getExploreLongUrl } from '../exploreUtils';
-import { t } from '../../locales';
+import CopyToClipboard from './CopyToClipboard';
+import { getShortUrl } from '../../utils/common';
+import { t } from '../locales';
 
 const propTypes = {
-  latestQueryFormData: PropTypes.object.isRequired,
+  url: PropTypes.string,
+  emailSubject: PropTypes.string,
```
Comment (mistercrunch) [inline superset/assets/javascripts/components/URLShortLinkButton.jsx]: Now that `emailSubject` and `emailContent` are the props, we should pass them good values where the components are used. Also always is `.isRequired` OR set `defaultProps`

Comment (ttannis) [reply superset/assets/javascripts/components/URLShortLinkButton.jsx]: 👍

**Comment by hughhhh** [comment]
@ttannis bumping for rebase!

File: superset/assets/javascripts/explore/components/ExploreViewContainer.jsx [hunk 5]
```diff
@@ -157,27 +151,6 @@ class ExploreViewContainer extends React.Component {
     }
   }
 
-  addHistory({ isReplace = false, title }) {
```
Comment (graceguo-supercat) [inline superset/assets/javascripts/explore/components/ExploreViewContainer.jsx]: if you removed this function, when user playing with chart controls, and they refresh their page (for any reason, without saving slice), will they lost the current query state? not all the query parameters are saved in slice yet.

Comment (ttannis) [reply superset/assets/javascripts/explore/components/ExploreViewContainer.jsx]: They will. I discussed this with @mistercrunch a couple weeks back and at that point we were okay with this side effect for a more consistent redirect implementation.

Comment (graceguo-supercat) [reply superset/assets/javascripts/explore/components/ExploreViewContainer.jsx]: In airbnb many users didn't realize there is a share linked button. they still get used to share their work by directly copy-n-paste url from browser. i guess this change will be dangerous to them :)

Comment (ttannis) [reply superset/assets/javascripts/explore/components/ExploreViewContainer.jsx]: @graceguo-supercat would you like to revisit/or discuss that decision again?

Comment (mistercrunch) [reply superset/assets/javascripts/explore/components/ExploreViewContainer.jsx]: @ttannis can we just keep `addHistory` as untouched as possible?

Comment (graceguo-supercat) [reply superset/assets/javascripts/explore/components/ExploreViewContainer.jsx]: i feel it's safe to keep query state in browser history.
- we still need to keep url scheme backward compatible right? Even we display url in browser like /r/123, when user request with old scheme, we should handle it correctly.
- after query state added into browser history, there is no real long url request happened unless user refreshed browser or someone accessed the shared url. So i guess the rate of old url scheme is pretty low.

**Review by mistercrunch — APPROVED** [review]

File: superset/assets/src/components/URLShortLinkButton.jsx [hunk 6]
```diff
@@ -25,20 +26,19 @@ export default class URLShortLinkButton extends React.Component {
...
 
   getCopyUrl() {
-    const longUrl = getExploreLongUrl(this.props.latestQueryFormData);
-    getShortUrl(longUrl, this.onShortUrlSuccess.bind(this));
+    getShortUrl(this.props.url, this.onShortUrlSuccess.bind(this));
   }
 
   renderPopover() {
-    const emailBody = t('Check out this slice: %s', this.state.shortUrl);
+    const emailBody = t('%s%s', this.emailContent, this.state.shortUrl);
```
Comment (mistercrunch) [inline superset/assets/src/components/URLShortLinkButton.jsx]: Shouldn't this be `this.props.emailContent`?

File: superset/assets/src/components/URLShortLinkButton.jsx [hunk 7]
```diff
@@ -25,20 +26,19 @@ export default class URLShortLinkButton extends React.Component {
...
     return (
       <Popover id="shorturl-popover">
         <CopyToClipboard
           text={this.state.shortUrl}
           copyNode={<i className="fa fa-clipboard" title={t('Copy to clipboard')} />}
         />
         &nbsp;&nbsp;
-        <a href={`mailto:?Subject=Superset%20Slice%20&Body=${emailBody}`}>
+        <a href={`mailto:?Subject=${this.emailSubject}%20&Body=${emailBody}`}>
```
Comment (mistercrunch) [inline superset/assets/src/components/URLShortLinkButton.jsx]: Shouldn't this be `this.props.emailContent`?

Comment (mistercrunch) [reply superset/assets/src/components/URLShortLinkButton.jsx]: @ttannis ^

**Comment by hughhhh** [comment]
🚢

**Pull request merged**